### PR TITLE
hashing tuples

### PR DESCRIPTION
**Pull request by gdkrmr**
Hashing for `NTuple{N, UInt8}`

had to change `Array{UInt8, 1}` to `Array{UInt8, N}` in `sha*` and `update!` for some reason, else I get the following error:
```
ERROR: LoadError: MethodError: no method matching sha1(::Array{UInt8,1})
Closest candidates are:
  sha1(::Union{Array{UInt8,1}, Tuple{Vararg{UInt8,N}}}) where N at /home/gkraemer/.julia/v0.6/SHA/src/SHA.jl:42
  sha1(!Matched::AbstractString) at /home/gkraemer/.julia/v0.6/SHA/src/SHA.jl:48
  sha1(!Matched::IO) at /home/gkraemer/.julia/v0.6/SHA/src/SHA.jl:54
```

### CODE REVIEW

**Comment by gdkrmr** [comment]
is the julia 0.4 compatibility an issue?

**Comment by staticfloat** [comment]
Hi @gdkrmr,

I'm curious, what use case do you have for hashing `NTuple`s?

It may make sense to just remove the type specification completely and allow this function to be duck-typed.

**Comment by gdkrmr** [comment]
It is a lot of small objects I want to hash, so I store them as NTuples instead of Arrays. The overall memory layout is nicer if I store an Array of NTuples instead of an Array of Arrays.
I did not try simply removing the Type specification.

**Comment by staticfloat** [comment]
> I did not try simply removing the Type specification.

Let's try that and see if it addresses your problems; the only thing I want to make sure of is that `eltype(data) == UInt8`, because hashing arrays of Int64's or whatever won't work properly.  Maybe we can just put in an assert that `eltype(data) == UInt8`, and make it clear in the docs that the type of `data` must conform to this.

**Comment by gdkrmr** [comment]
ok, I can do that. One question, is the error I got when using `Array{UInt8, 1}` a bug or a feature?

**Comment by staticfloat** [comment]
It looks like a little bit of both; the problem here is that all type parameters have to be uniquely determined, which means mixing a parameterized and non-parameterized piece in a type Union doesn't work, because then the parameter `N` has nothing constraining it.

One possible workaround is to use something like `Union{Array{UInt8,1}, NTuple{TypeVar(:N,Any),UInt8}}`.  This essentially works around the dispatch framework specializing on `N`, so there's no problems with parameterization.

**Comment by gdkrmr** [comment]
removing type specification did not help.

File: test/runtests.jl [hunk 1]
```diff
@@ -224,6 +235,16 @@ for idx in 1:length(ctxs)
...
 end
 
+# test error if eltype of input is not UInt8
+for f in sha_funcs
+    try
+        f(UInt32[0x23467, 0x324775])
+        warn("Non-UInt8 Arrays should fail")
+        nerrors += 1
+    end
+end
```
Comment (staticfloat) [inline test/runtests.jl]: I really like this test, thank you.

**Comment by staticfloat** [comment]
The only thing crashing right now is the fact that `NTuple`'s have an inferior implementation in Julia 0.4, and trying to construct a 1000000 element-long `NTuple` on Julia 0.4 causes a stack overflow.

I suggest just creating a new test case; something like 1000 `b`'s or something.

**Comment by gdkrmr** [comment]
I just saw that this never got merged, should I attempt to fix the tests?

**Comment by staticfloat** [comment]
Yes, that would be great, thanks.

**Comment by gdkrmr** [comment]
should be ok now, I hope I did not mess anything up. The unnecessary type check could be removed thanks to the updated type system in julia 0.6

**Comment by staticfloat** [comment]
Excellent, thank you for this @gdkrmr.

**Pull request merged**